apps/prairielearn/src/ee/lib/lti13.ts

```diff
@@ -0,0 +1,211 @@
+export const Lti13ClaimSchema = z.object({
+  'https://purl.imsglobal.org/spec/lti/claim/message_type': z.literal('LtiResourceLinkRequest'),
+  'https://purl.imsglobal.org/spec/lti/claim/version': z.literal('1.3.0'),
+  'https://purl.imsglobal.org/spec/lti/claim/deployment_id': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/target_link_uri': z.string(),
+  'https://purl.imsglobal.org/spec/lti/claim/resource_link': z.object({
+    id: z.string(),
+    description: z.string().nullish(),
+    title: z.string().nullish(),
+  }),
+  // https://www.imsglobal.org/spec/security/v1p0/#tool-jwt
+  // https://www.imsglobal.org/spec/security/v1p0/#id-token
+  iss: z.string(),
+  aud: z.string(),
+  sub: z.string(),
+  exp: z.number(),
+  iat: z.number(),
+  azp: z.string().optional(),
+  nonce: z.string(),
+
+  given_name: z.string().optional(),
+  family_name: z.string().optional(),
+  name: z.string().optional(),
+  email: z.string().optional(),
+  locale: z.string().optional(),
+  // Could be more from OIDC Standard Claims
+  'https://purl.imsglobal.org/spec/lti/claim/roles': z.string().array(),
+
+  'https://purl.imsglobal.org/spec/lti/claim/context': z
+    .object({
+      id: z.string(),
+      type: z.string().array().nullish(),
+      label: z.string().nullish(),
+      title: z.string().nullish(),
+    })
+    .nullish(),
+
+  'https://purl.imsglobal.org/spec/lti/claim/tool_platform': z
+    .object({
+      guid: z.string().max(255),
+      name: z.string().optional(),
+      contact_email: z.string().optional(),
+      description: z.string().optional(),
+      url: z.string().optional(),
+      product_family_code: z.string().optional(),
+      version: z.string().optional(),
+    })
+    .nullish(),
+
+  'https://purl.imsglobal.org/spec/lti/claim/role_scope_mentor': z.string().array().nullish(),
+
+  'https://purl.imsglobal.org/spec/lti/claim/launch_presentation': z
+    .object({
+      document_target: z.string().optional(),
+      height: z.number().optional(),
+      width: z.number().optional(),
+      return_url: z.string().optional(),
+      locale: z.string().optional(),
+    })
+    .nullish(),
+
+  'https://purl.imsglobal.org/spec/lti/claim/lis': z.any().nullish(),
+  'https://purl.imsglobal.org/spec/lti/claim/custom': z.any().nullish(),
+
+  // https://www.imsglobal.org/spec/lti/v1p3#vendor-specific-extension-claims
+  // My development Canvas sends their own named extension as a top level property
+  // "https://www.instructure.com/placement": "course_navigation"
+});
+export type Lti13ClaimType = z.infer<typeof Lti13ClaimSchema>;
+
+export class Lti13Claim {
+  private claims: Lti13ClaimType;
+  private req: Request;
+  private valid = false;
+
+  constructor(req: Request) {
+    try {
+      this.claims = Lti13ClaimSchema.passthrough().parse(req.session.lti13_claims);
+    } catch {
+      throw error.make(403, 'LTI session invalid or timed out, please try logging in again.');
```

Comment:
Here and elsewhere in this PR, we should use `new HttpStatusError` instead. I need to remove `make` and friends from the `@prairielearn/error` API.

Reply:
I think I've found and replaced all of these.